partitions.go

```diff
@@ -0,0 +1,367 @@
+	}
+	b.WriteByte(']')
+
+	return b.String()
+}
+
+func (ps *PartitionSpec) LastAssignedFieldID() int {
+	if len(ps.fields) == 0 {
+		return partitionDataIDStart
+	}
+
+	id := ps.fields[0].FieldID
+	for _, f := range ps.fields[1:] {
+		if f.FieldID > id {
+			id = f.FieldID
+		}
+	}
+	return id
+}
+
+// PartitionType produces a struct of the partition spec.
+//
+// The partition fields should be optional:
+//   - All partition transforms are required to produce null if the input value
+//     is null. This can happen when the source column is optional.
+//   - Partition fields may be added later, in which case not all files would
+//     have the result field and it may be null.
+//
+// There is a case where we can guarantee that a partition field in the first
+// and only parittion spec that uses a required source column will never be
+// null, but it doesn't seem worth tracking this case.
+func (ps *PartitionSpec) PartitionType(schema *Schema) *StructType {
+	nestedFields := []NestedField{}
+	for _, field := range ps.fields {
+		sourceType, ok := schema.FindTypeByID(field.SourceID)
+		if !ok {
+			continue
+		}
+		resultType := field.Transform.ResultType(sourceType)
+		nestedFields = append(nestedFields, NestedField{
+			ID:       field.FieldID,
+			Name:     field.Name,
+			Type:     resultType,
+			Required: false,
+		})
+	}
+	return &StructType{FieldList: nestedFields}
+}
+
+// ParseTransform takes the string representation of a transform as
+// defined in the iceberg spec, and produces the appropriate Transform
+// object or an error if the string is not a valid transform string.
+func ParseTransform(s string) (Transform, error) {
+	switch {
+	case strings.HasPrefix(s, "bucket"):
+		matches := regexFromBrackets.FindStringSubmatch(s)
+		if len(matches) != 2 {
+			break
+		}
+
+		n, _ := strconv.Atoi(matches[1])
+		return BucketTransform{N: n}, nil
+	case strings.HasPrefix(s, "truncate"):
+		matches := regexFromBrackets.FindStringSubmatch(s)
+		if len(matches) != 2 {
+			break
+		}
+
+		n, _ := strconv.Atoi(matches[1])
+		return TruncateTransform{W: n}, nil
+	default:
+		switch s {
+		case "identity":
+			return IdentityTransform{}, nil
+		case "void":
+			return VoidTransform{}, nil
+		case "year":
+			return YearTransform{}, nil
+		case "month":
+			return MonthTransform{}, nil
+		case "day":
+			return DayTransform{}, nil
+		case "hour":
+			return HourTransform{}, nil
+		}
+	}
+
+	return nil, fmt.Errorf("%w: %s", ErrInvalidTransform, s)
+}
+
+// Transform is an interface for the various Transformation types
+// in partition specs. Currently they do not yet provide actual
+// transformation functions or implementation. That will come later as
+// data reading gets implemented.
+type Transform interface {
+	fmt.Stringer
+	encoding.TextMarshaler
+	ResultType(t Type) Type
+}
+
+// IdentityTransform uses the identity function, performing no transformation
+// but instead partitioning on the value itself.
+type IdentityTransform struct{}
```

Comment:
Any reason why _transforms_ code is clubbed with _partitioning_ ?

Reply:
mostly because the `PartitionField` has a `Transform` object inside of it. Anything else that interacts with transforms, at least currently, only does so by first going through the partition spec and partition fields. I'm fine with moving the Transform code to it's own `transform.go` file if you think that would make more sense

Reply:
I think splitting the `transforms` into a separate `transform.go` makes sense

Reply:
updated and added.